File: backend/src/main/java/org/luxons/sevenwonders/controllers/GameBrowserController.java
package org.luxons.sevenwonders.controllers;

import java.security.Principal;
import java.util.Collection;
import java.util.Collections;

import org.hildan.livedoc.core.annotations.Api;
import org.luxons.sevenwonders.actions.CreateGameAction;
import org.luxons.sevenwonders.actions.JoinGameAction;
import org.luxons.sevenwonders.errors.ApiMisuseException;
import org.luxons.sevenwonders.lobby.Lobby;
import org.luxons.sevenwonders.lobby.Player;
import org.luxons.sevenwonders.repositories.LobbyRepository;
import org.luxons.sevenwonders.repositories.PlayerRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.messaging.handler.annotation.MessageMapping;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.messaging.simp.annotation.SendToUser;
import org.springframework.messaging.simp.annotation.SubscribeMapping;
import org.springframework.stereotype.Controller;
import org.springframework.validation.annotation.Validated;

/**
 * This is the place where the player looks for a game.
 */
@Api(name = "GameBrowser")
@Controller
public class GameBrowserController {

    private static final Logger logger = LoggerFactory.getLogger(GameBrowserController.class);

    private final LobbyController lobbyController;

    private final LobbyRepository lobbyRepository;

    private final PlayerRepository playerRepository;

    private final SimpMessagingTemplate template;

    @Autowired
    public GameBrowserController(LobbyController lobbyController, LobbyRepository lobbyRepository,
            PlayerRepository playerRepository, SimpMessagingTemplate template) {
        this.lobbyController = lobbyController;
        this.lobbyRepository = lobbyRepository;
        this.playerRepository = playerRepository;
        this.template = template;
    }

    /**
     * Gets the created or updated games. The list of existing games is received on this topic at once upon
     * subscription, and then each time the list changes.
     *
     * @param principal
     *         the connected user's information
     *
     * @return the current list of {@link Lobby}s
     */
    @SubscribeMapping("/games") // prefix /topic not shown
    public Collection<Lobby> listGames(Principal principal) {
        logger.info("Player '{}' subscribed to /topic/games", principal.getName());
        return lobbyRepository.list();
    }

    /**
     * Creates a new {@link Lobby}.
     *
     * @param action
     *         the action to create the game
     * @param principal
     *         the connected user's information
     *
     * @return the newly created {@link Lobby}
     */
    @MessageMapping("/lobby/create")
    @SendToUser("/queue/lobby/joined")
    public Lobby createGame(@Validated CreateGameAction action, Principal principal) {
        checkThatUserIsNotInAGame(principal, "cannot create another game");

        Player gameOwner = playerRepository.find(principal.getName());
        Lobby lobby = lobbyRepository.create(action.getGameName(), gameOwner);

        logger.info("Game '{}' ({}) created by {} ({})", lobby.getName(), lobby.getId(), gameOwner.getDisplayName(),
                gameOwner.getUsername());

        // notify everyone that a new game exists
        template.convertAndSend("/topic/games", Collections.singletonList(lobby));
        return lobby;
    }

    /**
     * Joins an existing {@link Lobby}.
     *
     * @param action
     *         the action to join the game
     * @param principal
     *         the connected user's information
     *
     * @return the {@link Lobby} that has just been joined
     */
    @MessageMapping("/lobby/join")
    @SendToUser("/queue/lobby/joined")
    public Lobby joinGame(@Validated JoinGameAction action, Principal principal) {
        checkThatUserIsNotInAGame(principal, "cannot join another game");

        Lobby lobby = lobbyRepository.find(action.getGameId());
        Player newPlayer = playerRepository.find(principal.getName());
        lobby.addPlayer(newPlayer);

        logger.info("Player '{}' ({}) joined game {}", newPlayer.getDisplayName(), newPlayer.getUsername(),
                lobby.getName());
        lobbyController.sendLobbyUpdateToPlayers(lobby);
        return lobby;
    }

    private void checkThatUserIsNotInAGame(Principal principal, String impossibleActionDescription) {
        Lobby lobby = playerRepository.find(principal.getName()).getLobby();
        if (lobby != null) {
            throw new UserAlreadyInGameException(lobby.getName(), impossibleActionDescription);
        }
    }

    static class UserAlreadyInGameException extends ApiMisuseException {
        UserAlreadyInGameException(String gameName, String impossibleActionDescription) {
            super("Client already in game '" + gameName + "', " + impossibleActionDescription);
        }
    }
}

File: backend/src/main/kotlin/org/luxons/sevenwonders/controllers/GameBrowserController.kt
package org.luxons.sevenwonders.controllers

import org.hildan.livedoc.core.annotations.Api
import org.luxons.sevenwonders.actions.CreateGameAction
import org.luxons.sevenwonders.actions.JoinGameAction
import org.luxons.sevenwonders.errors.ApiMisuseException
import org.luxons.sevenwonders.lobby.Lobby
import org.luxons.sevenwonders.repositories.LobbyRepository
import org.luxons.sevenwonders.repositories.PlayerRepository
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.messaging.handler.annotation.MessageMapping
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.messaging.simp.annotation.SendToUser
import org.springframework.messaging.simp.annotation.SubscribeMapping
import org.springframework.stereotype.Controller
import org.springframework.validation.annotation.Validated
import java.security.Principal

/**
 * This is the place where the player looks for a game.
 */
@Api(name = "GameBrowser")
@Controller
class GameBrowserController @Autowired constructor(
    private val lobbyController: LobbyController,
    private val lobbyRepository: LobbyRepository,
    private val playerRepository: PlayerRepository,
    private val template: SimpMessagingTemplate
) {

    /**
     * Gets the created or updated games. The list of existing games is received on this topic at once upon
     * subscription, and then each time the list changes.
     *
     * @param principal the connected user's information
     *
     * @return the current list of [Lobby]s
     */
    @SubscribeMapping("/games") // prefix /topic not shown
    fun listGames(principal: Principal): Collection<Lobby> {
        logger.info("Player '{}' subscribed to /topic/games", principal.name)
        return lobbyRepository.list()
    }

    /**
     * Creates a new [Lobby].
     *
     * @param action the action to create the game
     * @param principal the connected user's information
     *
     * @return the newly created [Lobby]
     */
    @MessageMapping("/lobby/create")
    @SendToUser("/queue/lobby/joined")
    fun createGame(@Validated action: CreateGameAction, principal: Principal): Lobby {
        checkThatUserIsNotInAGame(principal, "cannot create another game")

        val gameOwner = playerRepository.find(principal.name)
        val lobby = lobbyRepository.create(action.gameName, gameOwner)

        logger.info(
            "Game '{}' ({}) created by {} ({})", lobby.name, lobby.id, gameOwner.displayName, gameOwner.username
        )

        // notify everyone that a new game exists
        template.convertAndSend("/topic/games", listOf(lobby))
        return lobby
    }

    /**
     * Joins an existing [Lobby].
     *
     * @param action the action to join the game
     * @param principal the connected user's information
     *
     * @return the [Lobby] that has just been joined
     */
    @MessageMapping("/lobby/join")
    @SendToUser("/queue/lobby/joined")
    fun joinGame(@Validated action: JoinGameAction, principal: Principal): Lobby {
        checkThatUserIsNotInAGame(principal, "cannot join another game")

        val lobby = lobbyRepository.find(action.gameId!!)
        val newPlayer = playerRepository.find(principal.name)
        lobby.addPlayer(newPlayer)

        logger.info(
            "Player '{}' ({}) joined game {}", newPlayer.displayName, newPlayer.username, lobby.name
        )
        lobbyController.sendLobbyUpdateToPlayers(lobby)
        return lobby
    }

    private fun checkThatUserIsNotInAGame(principal: Principal, impossibleActionDescription: String) {
        val player = playerRepository.find(principal.name)
        if (player.isInLobby || player.isInGame) {
            throw UserAlreadyInGameException(player.lobby.name, impossibleActionDescription)
        }
    }

    internal class UserAlreadyInGameException(gameName: String, impossibleActionDescription: String) :
        ApiMisuseException("Client already in game '$gameName', $impossibleActionDescription")

    companion object {

        private val logger = LoggerFactory.getLogger(GameBrowserController::class.java)
    }
}
